### PR TITLE
Extract crypto workers from core

### PR DESCRIPTION
**Pull request by inorick**
Extract workers from core so they can be driven by individual tasks. The core and every worker only sees sources and sinks to communicate with other parts safely. This is a precondition to make the code async.

### CODE REVIEW

File: heimlig/src/common/jobs.rs [hunk 1]
```diff
@@ -1,21 +1,33 @@
...
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum RequestType {
+    ImportKey,
+    GetRandom,
+    EncryptChaChaPoly,
+    EncryptChaChaPolyExternalKey,
+    DecryptChaChaPoly,
+    DecryptChaChaPolyExternalKey,
+}
+
```
Comment (SCingolani) [inline heimlig/src/common/jobs.rs:25]: I have found this issue before, i.e., having to declare a separate "simple" enum to represent the possible variants of a proper sum type enum. I wonder if there's a crate that implements a macro for this?

File: heimlig/src/client/api.rs [hunk 2]
```diff
@@ -1,52 +1,48 @@
...
-    requests_sink: Req,
-    responses_source: Resp,
+    requests: Req,
+    responses: Resp,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum Error {
+    Queue(queues::Error),
 }
```
Comment (SCingolani) [inline heimlig/src/client/api.rs:14]: Do we see other errors being necessary here? If not we could do `pub use queues::Error` or so to re-export the error?

File: heimlig/src/hsm/core.rs [hunk 3]
```diff
@@ -1,203 +1,202 @@
...
+    'keystore,
+    M: RawMutex,
+    ReqSrc: Iterator<Item = (usize, Request<'data>)>,
+    RespSink: ResponseSink<'data>,
+    ReqSink: RequestSink<'data>,
+    RespSrc: Iterator<Item = (usize, Response<'data>)>,
+    const MAX_REQUEST_TYPES: usize = 8,
+    const MAX_WORKERS: usize = 8,
+> {
+    key_store: Option<&'keystore Mutex<M, &'keystore mut (dyn KeyStore + Send)>>,
```
[the remark this answers is not in the document]
Comment (SCingolani) [reply heimlig/src/hsm/core.rs:30]: Already discussed offline, in the end paying a small runtime price in order to not have confusing errors where a Core without a KeyStore would actually report KeyNotFound as an error is acceptable.

File: heimlig/src/hsm/core.rs [hunk 4]
```diff
@@ -1,203 +1,202 @@
...
-    use heapless::spsc::{Consumer, Producer, Queue};
-
-    const QUEUE_SIZE: usize = 8;
 
-    struct RequestQueueSource<'ch, 'a> {
-        consumer: Consumer<'ch, Request<'a>, QUEUE_SIZE>,
+    pub fn execute(&mut self) -> Result<(), Error> {
+        self.process_worker_responses()?;
+        self.process_client_requests()?;
+        self.process_worker_responses()?;
```
[the remark this answers is not in the document]
Comment (inorick) [reply heimlig/src/hsm/core.rs]: I removed the second check.

File: heimlig/src/hsm/workers/chachapoly_worker.rs [hunk 5]
```diff
@@ -1,8 +1,117 @@
...
 
-pub struct ChachaPolyWorker;
+pub struct ChaChaPolyWorker<
+    'data,
+    'keystore,
+    M: RawMutex,
+    ReqSrc: Iterator<Item = (usize, Request<'data>)>,
+    RespSink: ResponseSink<'data>,
+> {
+    pub key_store: Option<&'keystore Mutex<M, &'keystore mut (dyn KeyStore + Send)>>,
```
[the remark this answers is not in the document]
Comment (inorick) [reply heimlig/src/hsm/workers/chachapoly_worker.rs:19]: Discussed offline. See above.

**Pull request merged**